server/src/main/java/org/elasticsearch/rest/action/DispatchingRestToXContentListener.java

```diff
@@ -25,7 +24,7 @@
  * Response listener for REST requests which dispatches the serialization of the response off of the thread on which the response was
  * received, since that thread is often a transport thread and XContent serialization might be expensive.
  */
-public class DispatchingRestToXContentListener<Response extends ToXContentObject> extends RestActionListener<Response> {
+public class DispatchingRestToXContentListener<Response extends StatusToXContentObject> extends RestActionListener<Response> {
```

Comment:
In this branch `RestGetMappingAction` returns a status code based on the response, so I've decided to require a `StatusToXContentObject` in this helper class.

Reply:
Should we forward port this change to `master`?

Reply:
Seems reasonable yes. I'm working through the other changes here too, but in addition how about this rather than introducing a special case specifically for `IndicesSegmentResponse`:

```diff
diff --git a/server/src/main/java/org/elasticsearch/common/xcontent/StatusToXContentObject.java b/server/src/main/java/org/elasticsearch/common/xcontent/StatusToXContentObject.java
index b75f69bd286..d599461f862 100644
--- a/server/src/main/java/org/elasticsearch/common/xcontent/StatusToXContentObject.java
+++ b/server/src/main/java/org/elasticsearch/common/xcontent/StatusToXContentObject.java
@@ -9,6 +9,8 @@ package org.elasticsearch.common.xcontent;

 import org.elasticsearch.rest.RestStatus;

+import java.io.IOException;
+
 /**
  * Objects that can both render themselves in as json/yaml/etc and can provide a {@link RestStatus} for their response. Usually should be
  * implemented by top level responses sent back to users from REST endpoints.
@@ -19,4 +21,18 @@ public interface StatusToXContentObject extends ToXContentObject {
      * Returns the REST status to make sure it is returned correctly
      */
     RestStatus status();
+
+    static StatusToXContentObject withStatus(RestStatus status, ToXContentObject xContent) {
+        return new StatusToXContentObject() {
+            @Override
+            public RestStatus status() {
+                return status;
+            }
+
+            @Override
+            public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+                return xContent.toXContent(builder, params);
+            }
+        };
+    }
 }
diff --git a/server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestIndicesSegmentsAction.java b/server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestIndicesSegmentsAction.java
index 06b8959f399..e6ea63b0764 100644
--- a/server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestIndicesSegmentsAction.java
+++ b/server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestIndicesSegmentsAction.java
@@ -12,8 +12,10 @@ import org.elasticsearch.action.admin.indices.segments.IndicesSegmentsRequest;
 import org.elasticsearch.action.support.IndicesOptions;
 import org.elasticsearch.client.node.NodeClient;
 import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.xcontent.StatusToXContentObject;
 import org.elasticsearch.rest.BaseRestHandler;
 import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.RestStatus;
 import org.elasticsearch.rest.action.DispatchingRestToXContentListener;
 import org.elasticsearch.rest.action.RestCancellableNodeClient;
 import org.elasticsearch.threadpool.ThreadPool;
@@ -53,6 +55,7 @@ public class RestIndicesSegmentsAction extends BaseRestHandler {
         indicesSegmentsRequest.indicesOptions(IndicesOptions.fromRequest(request, indicesSegmentsRequest.indicesOptions()));
         return channel -> new RestCancellableNodeClient(client, request.getHttpChannel()).admin().indices().segments(
                 indicesSegmentsRequest,
-                new DispatchingRestToXContentListener<>(threadPool.executor(ThreadPool.Names.MANAGEMENT), channel, request));
+                new DispatchingRestToXContentListener<>(threadPool.executor(ThreadPool.Names.MANAGEMENT), channel, request)
+                        .map(r -> StatusToXContentObject.withStatus(RestStatus.OK, r)));
     }
 }
```